### PR TITLE
[api] improve config options

### PR DESCRIPTION
**Pull request by ashulin**
## Purpose of this pull request

 - [x] Provides more friendly options capabilities.
 - [x] Add validation rules for options, and used for Web-UI to prompt user to configure option value;

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    public static ReadonlyConfig fromMap(Map<String, Object> map) {
+        return new ReadonlyConfig(treeMap(map));
+    }
+
+    public static ReadonlyConfig fromConfig(Config config) {
+        try {
+            return fromMap(JACKSON_MAPPER.readValue(
+                config.root().render(ConfigRenderOptions.concise()),
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java:57]: I think use jackson to deserialize config to `ReadonlyConfig` not a cool way. Why just iterate over all values in config, and put into map?

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    public <T> T get(Option<T> option) {
+        return getOptional(option).orElseGet(option::defaultValue);
+    }
+
+    public Map<String, String> toMap() {
+        synchronized (this.confData) {
+            Map<String, Object> flatteningMap = flatteningMap(confData);
+            Map<String, String> result = new HashMap<>(flatteningMap.size());
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java]: `flatteningMap.size()` will convert to array size in map, not map can store data size. If you want set size in hashmap, please use 
```suggestion
            Map<String, String> result = new HashMap<>((float)flatteningMap.size()/0.75 + 1);
```
Or just remove set capacity (recommend).

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T> Optional<T> getOptional(Option<T> option) {
+        if (option == null) {
+            throw new NullPointerException("Option not be null.");
+        }
+        String[] keys = option.key().split("\\.");
+        synchronized (this.confData) {
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java]: getOptional use `synchronized` not a great idea. The performance must not good. Maybe you can use cache or something to precompute

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        } else if (obj instanceof ReadonlyConfig) {
+            Map<String, Object> otherConf = ((ReadonlyConfig) obj).confData;
+
+            for (Map.Entry<String, Object> e : this.confData.entrySet()) {
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java]: Just use `map.equals` can solve the requirement.

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java (hunk 1 above)
Comment (ashulin) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java:57]: Here the `typesafe config` is converted to a json string and parsed into the required map using jackson

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java (hunk 2 above)
Comment (ashulin) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java]: I don't understand, this method doesn't change the size of map keys.

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java (hunk 3 above)
Comment (ashulin) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java]: You have point here.

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java (hunk 1 above)
Comment (Hisoka-X) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java:57]: > Why just iterate over all values in config, and put into map?

My fault, It should be `Why not just iterate over all values in config, and put into map`. If do that, we wouldn't use jackson to serialize and deserialize

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java (hunk 2 above)
Comment (Hisoka-X) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java]: Check this: https://cloud.tencent.com/developer/article/1913508

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java (hunk 1 above)
Comment (ashulin) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java:57]: I have tried to convert `config` without using jackson, but the `config` is more complicated, and there will be many problems in my own implementation. config provides `config.root().render(ConfigRenderOptions.concise())` for converting to json, and it has been well tested.

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java (hunk 2 above)
Comment (ashulin) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java]: You're absolutely right. Thanks

**Comment by ashulin** [comment]
@Hisoka-X PTAL

**Review by Hisoka-X — APPROVED** [review]
LGTM

**Pull request merged**